lib/formz.dart

```diff
@@ -166,5 +198,179 @@ mixin FormzMixin {
+  ///  if (alreadyExist) {
+  ///    return const EmailValidationError.alreadyExists();
+  ///  }
+  ///
+  ///  return null;
+  ///}
+  /// ```
+  Future<TError?> validate(TInput input);
+
+  /// Returns whether the [TInput] is ready for validation.
+  ///
+  /// ```dart
+  /// @override
+  ///bool canValidate(Email input) {
+  ///  if (input.validationStatus.isValidated) {
+  ///    return true;
+  ///  }
+  ///  final validFormat = RegVal.hasMatch(
+  ///    input.value,
+  ///    _kEmailPattern,
+  ///  );
+  ///  return validFormat;
+  ///}
+  /// ```
+  bool canValidate(TInput input) => true;
+}
+
+/// Enum representing the validation status of a [AsyncFormzInput] object.
+enum AsyncFormzInputValidationStatus {
+  /// Indicates whether the [AsyncFormzInput] has not been validated.
+  pure,
+
+  /// Indicates whether the [AsyncFormzInput] is being validated.
+  validating,
+
+  /// Indicates whether the [AsyncFormzInput] has been validated.
+  validated,
+}
+
+/// Useful extensions on [AsyncFormzInputValidationStatus]
+extension AsyncFormzInputValidationStatusX on AsyncFormzInputValidationStatus {
+  /// Indicates whether the [AsyncFormzInput] has not been validated.
+  bool get isPure => this == AsyncFormzInputValidationStatus.pure;
+
+  /// Indicates whether the [AsyncFormzInput] is being validated.
+  bool get isValidating => this == AsyncFormzInputValidationStatus.validating;
+
+  /// Indicates whether the [AsyncFormzInput] is not being validated.
+  bool get isNotValidating => !isValidating;
+
+  /// Indicates whether the [AsyncFormzInput] is validated.
+  bool get isValidated => this == AsyncFormzInputValidationStatus.validated;
+}
+
+/// {@template form_input}
+/// A [AsyncFormzInput] represents the value of a single form input field.
+/// It contains information about the [value] as well as validity.
+///
+/// [AsyncFormzInput] should be extended to define custom [AsyncFormzInput]
+/// instances.
+///
+/// ```dart
+///class Email extends AsyncFormzInput<String, EmailValidationError> {
+///  const Email(
+///    super.value, {
+///    super.error,
+///    super.validationStatus,
+///    this.isRequired = true,
+///  });
+///
+///  final bool isRequired;
+///}
+/// ```
+/// {@endtemplate}
+@immutable
+abstract class AsyncFormzInput<T, E> implements FormzInputBase<T, E> {
+  /// Constructor which create a [FormzInput] with a given value.
+  const AsyncFormzInput(
+    this.value, {
+    this.error,
+    AsyncFormzInputValidationStatus? validationStatus,
+  }) : validationStatus =
+            validationStatus ?? AsyncFormzInputValidationStatus.pure;
+
+  @override
+  final T value;
+
+  @override
+  bool get isValid => validationStatus.isValidated && error == null;
+
+  @override
+  bool get isNotValid => !isValid;
+
+  @override
+  final E? error;
+
+  /// The validation status of the [AsyncFormzInput].
+  final AsyncFormzInputValidationStatus validationStatus;
```

Comment:
With the validator being independent of the input class, there is no safeguard (and out-of-the-box clarity) on who should change this value. 

The package usage becomes too much dependent on the boilerplate to work consistently.